sw/airborne/subsystems/ins/ins_int.c

```diff
@@ -435,33 +466,37 @@ void ins_int_update_gps(struct GpsState *gps_s)
 void ins_int_update_gps(struct GpsState *gps_s __attribute__((unused))) {}
 #endif /* USE_GPS */
 
-
+/** agl_cb
+ * This callback handles all estimates of the height of the vehicle above the ground under it
+ * This requires that the barometer is defined and used
+ */
+static void agl_cb(uint8_t __attribute__((unused)) sender_id, __attribute__((unused)) float distance) {
+#if USE_BARO_BOARD && USE_VFF_EXTENDED
+  if (sender_id == AGL_SONAR_ADC_ID ||
+      sender_id == AGL_SONAR_ARDRONE2_ID ||
+      sender_id == AGL_SONAR_NPS_ID ||
+      sender_id == AGL_SONAR_PX4FLOW_ID){
 #if USE_SONAR
-static void sonar_cb(uint8_t __attribute__((unused)) sender_id, float distance)
-{
-  static float last_offset = 0.;
-
-  /* update filter assuming a flat ground */
-  if (distance < INS_SONAR_MAX_RANGE && distance > INS_SONAR_MIN_RANGE
+    if (distance > 0
+        && vff.z > -INS_SONAR_MAX_RANGE
+        && vff.z < -INS_SONAR_MIN_RANGE
```

Comment:
So I am reviving this PR but have a couple of quips. Firstly, This line basically assumes that the vertical filter is converged and close to the true value but what if it isn't... What if the sonar is measuring the correct distance but the vertical filter is not correct? 

What is a better way to test that the measured sonar value is correct?